src/com/sun/jna/internal/Cleaner.java

```diff
@@ -44,65 +50,79 @@ public static Cleaner getCleaner() {
         return INSTANCE;
     }
 
+    // Guard for trackedObjects and cleanerThread. The readlock is utilized when
+    // the trackedObjects are manipulated, the writelock protectes starting and
+    // stopping the CleanerThread
+    private final ReadWriteLock cleanerThreadLock = new ReentrantReadWriteLock();
```

Comment:
I believe the readlock is not necessary at all. `trackedObjects` is an `AtomicLong` and doesn't need any protection.
The important step is to protect the starting and stopping of the CleanerThread. In particular, it must be ensured that the thread doesn't terminate while a new reference is added without starting a new CleanerThread..

Reply:
The `cleanerThreadLock` protects the interaction of `trackedObjects` and `cleanerThread`. The critical section is the starting _and_ stopping of the Thead if and only iff the number of tracked objects reaches zero or is above 1.

This is the sequence I want to prevent:

1. `T1` using the cleaner calls `register` and runs _before_ `incrementAndGet` is invoked
2. `CleanerThread` reaches the `trackedObjects.get() == 0` check and enters the `if` block, but does not execute further yet.
3. `T1` executes `incrementAndGet` and receives `1`, checks the value of `cleanerThread` and finds a `Thread` there.
4. `CleanerThread` continues, clears `cleanerThread` and finishes execution

There is a subtle issue there, the `break` needs to move into the inner `if` in line 173 (https://github.com/java-native-access/jna/pull/1617/files/b901c180f4a1026675a2859bbf24727574e19e6a#diff-11b3b0981223743529d972f317f26dfd9488b4a9b80a34a2d6f76cc2c58dcc00R173) into the inner if.

Reply:
Yes. I think that can be done with a single lock, i.e. what's currently the write lock.

Pseudocode:
```
Object lock = new Object();

add() {
trackedObjects.increment();
if (cleanerThread == null) {
     synchronized (lock) {
         if (cleanerThread == null) { cleanerThread = new CleanerThread(); cleanerThread.start(); }
     }
}

class CleanerThread {
  run() {
     //...
     if (trackedObjects.get() == 0) {
         synchronized (lock) {
             if (trackedObjects.get() == 0) { cleanerThread = null; break; }
         }
     }
  }
}
```